### ISSUE TITLE
False negative R0204 on async functions

### ISSUE THREAD

**Issue by rudineirk** [PyCQA/pylint#1279]
### Steps to reproduce
With async functions, pylint is returning false negative for variable type redefinition (R0204)

Some examples that cause the error:
```python
async def test_a():
    data = [
        {'test': 1},
        {'test': 2},
    ]
    return data

async def test_b():
    data = {'test': 1}
    return data
```

```python
class Test:
    async def test_a(self):
        data = [
            {'test': 1},
            {'test': 2},
        ]
        return data

    async def test_b(self):
        data = {'test': 1}
        return data
```

### Current behavior
```
R:  7, 4: Redefinition of data type from list to dict (redefined-variable-type)
```

### Expected behavior
no type redefinition messages

### pylint --version output
pylint 1.6.4, 
astroid 1.4.9
Python 3.5.2 (default, Sep 14 2016, 11:28:32) 
[GCC 6.2.1 20160901 (Red Hat 6.2.1-1)]

**Comment by PCManticore** [comment]
``redefined-variable-type`` is deprecated (actually moved into an extension, thus needing explicit activation). Please ignore it for now, since it has lots of edge cases.

**Comment by PCManticore** [comment]
Forgot to mention that this is in the latest code from the master branch, not yet released in the wild.
